devicesetup/src/main/java/io/particle/android/sdk/accountsetup/TwoFactorActivity.java

```diff
@@ -0,0 +1,134 @@
+package io.particle.android.sdk.accountsetup;
+
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.text.TextUtils;
+import android.widget.EditText;
+
+import javax.inject.Inject;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+import io.particle.android.sdk.cloud.ParticleCloud;
+import io.particle.android.sdk.cloud.SDKGlobals;
+import io.particle.android.sdk.cloud.exceptions.ParticleCloudException;
+import io.particle.android.sdk.devicesetup.ParticleDeviceSetupLibrary;
+import io.particle.android.sdk.devicesetup.R;
+import io.particle.android.sdk.devicesetup.R2;
+import io.particle.android.sdk.di.ApModule;
+import io.particle.android.sdk.ui.BaseActivity;
+import io.particle.android.sdk.ui.NextActivitySelector;
+import io.particle.android.sdk.utils.Async;
+import io.particle.android.sdk.utils.SEGAnalytics;
+import io.particle.android.sdk.utils.TLog;
+import io.particle.android.sdk.utils.ui.ParticleUi;
+import io.particle.android.sdk.utils.ui.Ui;
+
+public class TwoFactorActivity extends BaseActivity {
+    public static final String EMAIL_EXTRA = "email";
+    public static final String PASSWORD_EXTRA = "password";
+    public static final String MFA_EXTRA = "mfa_token";
+
+    private static final TLog log = TLog.get(TwoFactorActivity.class);
+
+    /**
+     * Keep track of the login task to ensure we can cancel it if requested, ensure against
+     * duplicate requests, etc.
+     */
+    private Async.AsyncApiWorker<ParticleCloud, Void> loginTask = null;
+
+    @Inject
+    protected ParticleCloud sparkCloud;
+
+    @BindView(R2.id.verificationCode)
+    protected EditText verificationCode;
+
+    @OnClick(R2.id.recover_auth)
+    public void onRecover() {
+        Intent intent = new Intent(Intent.ACTION_VIEW);
+        intent.setData(Uri.parse(getString(R.string.recovery_link)));
+        startActivity(intent);
+    }
+
+    @OnClick(R2.id.action_verify)
+    public void onVerify() {
+        if (loginTask != null) {
+            log.wtf("Login being attempted again even though the button isn't enabled?!");
+            return;
+        }
+
+        final String code = verificationCode.getText().toString();
+
+        if (TextUtils.isEmpty(code)) {
+            verificationCode.setError(getString(R.string.error_field_required));
+            verificationCode.requestFocus();
+        } else {
+            Intent callingIntent = getIntent();
+            login(callingIntent.getStringExtra(EMAIL_EXTRA), callingIntent.getStringExtra(PASSWORD_EXTRA),
+                    callingIntent.getStringExtra(MFA_EXTRA), code);
+        }
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_two_factor);
+        ParticleDeviceSetupLibrary.getInstance().getApplicationComponent().activityComponentBuilder()
```

Comment:
bro, do you even line break?  ;)

Again, not blocking PR for this, but in the future, let's use builder-style wrapping for this kind of long call chain.